README.md
# DateTime Pattern Placeholder

[![Latest version](https://img.shields.io/npm/v/datetime-placeholder)
 ![Dependency status](https://img.shields.io/librariesio/release/npm/datetime-placeholder)
](https://www.npmjs.com/package/datetime-placeholder)
[![Coverage](https://codecov.io/gh/prantlf/datetime-placeholder/branch/master/graph/badge.svg)](https://codecov.io/gh/prantlf/datetime-placeholder)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/9f1034029c0747a980cd49f64f16338b)](https://www.codacy.com/app/prantlf/datetime-placeholder?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=prantlf/datetime-placeholder&amp;utm_campaign=Badge_Grade)

Simplifies a date/time-formatting pattern using [Unicode LDML] tokens to a pattern usable in date/time pickers.

Although [Unicode LDML] is a widely accepted standard for tokens in date/time formatting patterns, the variety of the tokens is not comprehensible, when presented in a date/time picker to an end-user. Patterns presented as value placeholders can be simpler, because the parser accepts more variants of the input form of a date/value part.

| Locale | Formatting pattern  | Picker placeholder          |
|:-------|:--------------------|:----------------------------|
| en     | M/d/yy, h:mm a      | mm/dd/yy, hh:mm aa          |
| cs     | d. MMMM y H:mm:ss z | dd. mmm yyyy hh:mm:ss zz    |

```js
const placeholder = simplifyDateTimePattern('d.M.yy') // dd.mm.yy
```

* ES, CJS and UMD module exports.
* TypeScript type declarations (typings).
* No other dependencies.
* Tiny code base - 1.88 kB minified, 801 B gzipped, 731 B brotlied.

Related projects:

* [date-and-time-formatter] - formats a date/time value to a localised string using a pattern consisting of [Unicode LDML] tokens.
* [datetime-locale-patterns] - provides localized date/time format patterns for styles `full`, `long`, `medium` and `short` using [Unicode CLDR], compliant with [Unicode LDML].
* [intl-datetimeformat-options] - provides localized date/time format patterns for styles `full`, `long`, `medium` and `short`, using [`Intl.DateTimeFormat`].
* [intl-datetimeformat-pattern] - creates a valid [`Intl.DateTimeFormat`] options object from a [Unicode CLDR] skeleton or token pattern.

## Installation

This module can be installed in your project using [NPM], [PNPM] or [Yarn]. Make sure, that you use [Node.js] version 16.14 or newer.

```sh
$ npm i datetime-placeholder
$ pnpm i datetime-placeholder
$ yarn add datetime-placeholder
```

Functions are exposed as named exports from ES and CJS modules, for example:

```js
import { simplifyDateTimePattern } from 'datetime-placeholder'
```

```js
const { simplifyDateTimePattern } = require('datetime-placeholder')
```

A UMD module can be loaded to the browser either directly:

```html
<script src="https://unpkg.com/datetime-placeholder@1.0.0/lib/index.min.js"></script>
<script>
  const { simplifyDateTimePattern } = window.dateTimePlaceholder
</script>
```

Or using an AMD module loader:

```html
<script>
  require([
    'https://unpkg.com/datetime-placeholder@1.0.0/lib/index.min.js'
  ], ({ simplifyDateTimePattern }) => {
    ...
  })
</script>
```

## API

### simplifyDateTimePattern(pattern, options?)

Returns a simplified pattern to show to the end-user in a date/time picker.

The **pattern** argument must be a date/time-formatting pattern using [Unicode LDML] tokens.

```js
import { simplifyDateTimePattern } from 'datetime-placeholder'

const placeholder = simplifyDateTimePattern('M/d/yy')
console.log(placeholder) // prints 'mm/dd/yy'
```

### simplifyDateTimePatternToParts(pattern, options?)

Returns a simplified pattern to show to the end-user in a date/time picker as an array of parts, which values are supposed to be concatenated together.

The **pattern** argument must be a date/time-formatting pattern using [Unicode LDML] tokens.

```js
import { simplifyDateTimePatternToParts } from 'datetime-placeholder'

const placeholder = simplifyDateTimePatternToParts('M/d/yy')
console.log(placeholder) // prints the following:
// [{ type: 'month', value: 'mm',
//    type: 'literal', value: '/',
//    type; 'day', value: 'dd',
//    type: 'literal', value: '/',
//    type: 'year': value: 'yy' }]
```

### Options

The pattern simplification can be customised by an optional parameter - an object with the following properties:

| Name         | Default       | Description |
|:-------------|:--------------|:------------|
| `letterCase` | `'lowercase'` | if the pattern should be rendered lowercase (`'lowercase'`) or uppercase (`'uppercase'`) |


## Pattern Conversion

The characters wrapped between two single quotes characters (`'`) are escaped.
Two single quotes in a row, whether inside or outside a quoted sequence,
represent a "real" single quote. (see the last example)

The input pattern is based on [Unicode Technical Standard #35].
Only those tokens are supported, which can appear in patterns for formatting
instances of `Date` by [`Intl.DateTimeFormat`]. For example, only the "formatting"
variant of a token (M, E) is supported, not the "stand-alone" one (L, i).
("Formatting" means declined according to the rules of the language
 in the context of a date. "Stand-alone" means always nominative singular.)

Accepted patterns:

| Unit                         | Pattern | Placeholder | Meaning      |
|------------------------------|---------|-------------|--------------|
| Era                          | G...    | ee          | era          |
| Calendar year                | y       | yyyy        | full year    |
|                              | yy      | yy          | 2-digit year |
|                              | yyy...  | yyyy        | full year    |
| Month                        | M | MM  | mm          | month number |
|                              | MMM...  | mmm         | month name   |
| Day of month                 | d | dd  | dd          | day number   |
| Day of week                  | E...    | www         | weekday name |
| AM, PM                       | a...    | aa          | AM/PM        |
| AM, PM, noon, midnight       | b...    | aa          |              |
| Flexible day period          | B...    | aa          |              |
| Hour [1-12]                  | h | hh  | hh          | hours        |
| Hour [0-23]                  | H | HH  | hh          |              |
| Hour [0-11]                  | K | KK  | hh          |              |
| Hour [1-24]                  | k | kk  | hh          |              |
| Minute                       | m | mm  | mm          | minutes      |
| Second                       | s | ss  | ss          | seconds      |
| Fraction of second           | S...    | mmm         | milliseconds |
| Timezone (ISO-8601 w/ Z)     | X...    | zz          | time zone    |
| Timezone (ISO-8601 w/o Z)    | x...    | zz          |              |
| Timezone (GMT)               | O...    | zz          |              |
| Timezone (specific non-loc.) | z...    | zz          |              |

Although some placeholders are the same, their meaning is clear from the context within the pattern, because people know how to write down a date/time value.

### Examples

```js
// Represent 11 February 2014 in middle-endian format
simplifyDateTimePattern('MM/dd/yyyy')
//=> 'mm/dd/yyyy'
```

```js
// Escape string by single quote characters
simplifyDateTimePattern("h 'o''clock'", { letterCase: 'uppercase' })
//=> "HH o'clock"
```

## Contributing

In lieu of a formal styleguide, take care to maintain the existing coding style.  Add unit tests for any new or changed functionality. Lint and test your code using Grunt.

## License

Copyright (c) 2023 Ferdinand Prantl

Licensed under the MIT license.

[Node.js]: http://nodejs.org/
[NPM]: https://www.npmjs.com/
[PNPM]: https://pnpm.io/
[Yarn]: https://yarnpkg.com/
[date-and-time-formatter]: https://github.com/prantlf/date-and-time-formatter
[datetime-locale-patterns]: https://github.com/prantlf/datetime-locale-patterns
[intl-datetimeformat-pattern]: https://github.com/caridy/intl-datetimeformat-pattern
[intl-datetimeformat-options]: https://github.com/prantlf/intl-datetimeformat-options
[Unicode LDML]: https://unicode.org/reports/tr35/
[Unicode CLDR]: https://cldr.unicode.org/
[`Intl.DateTimeFormat`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat
[Unicode Technical Standard #35]: https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table


CHANGELOG.md
# [1.1.0](https://github.com/prantlf/datetime-placeholder/compare/v1.0.0...v1.1.0) (2023-05-16)


### Features

* Add an option for simplified uppercase patterns ([ca4ab01](https://github.com/prantlf/datetime-placeholder/commit/ca4ab01df5519e9b14ecf03432555552fc6f07a0))

# 1.0.0 (2023-03-19)
 
Initial release


test/parts.test.js
import tehanu from 'tehanu'
import { deepStrictEqual } from 'assert'
import { simplifyDateTimePatternToParts } from '../lib/index.js'

const test = tehanu(import.meta.url)

test('era', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('G'), [
    { type: 'era', value: 'ee' }
  ])
})

test('year', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('y'), [
    { type: 'year', value: 'yyyy' }
  ])
  deepStrictEqual(simplifyDateTimePatternToParts('yy'), [
    { type: 'year', value: 'yy' }
  ])
})

test('month', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('M'), [
    { type: 'month', value: 'mm' }
  ])
})

test('day of week', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('E'), [
    { type: 'weekday', value: 'www' }
  ])
})

test('day', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('d'), [
    { type: 'day', value: 'dd' }
  ])
})

test('hour', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('h'), [
    { type: 'hour', value: 'hh' }
  ])
})

test('minute', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('m'), [
    { type: 'minute', value: 'mm' }
  ])
})

test('second', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('s'), [
    { type: 'second', value: 'ss' }
  ])
})

test('fraction of second', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('S'), [
    { type: 'fractionalSecond', value: 'mmm' }
  ])
})

test('day period', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('a'), [
    { type: 'dayPeriod', value: 'aa' }
  ])
  deepStrictEqual(simplifyDateTimePatternToParts('b'), [
    { type: 'dayPeriod', value: 'aa' }
  ])
  deepStrictEqual(simplifyDateTimePatternToParts('B'), [
    { type: 'dayPeriod', value: 'aa' }
  ])
})

test('time zone', () => {
  deepStrictEqual(simplifyDateTimePatternToParts('Z'), [
    { type: 'timeZoneName', value: 'zz' }
  ])
  deepStrictEqual(simplifyDateTimePatternToParts('z'), [
    { type: 'timeZoneName', value: 'zz' }
  ])
  deepStrictEqual(simplifyDateTimePatternToParts('O'), [
    { type: 'timeZoneName', value: 'zz' }
  ])
  deepStrictEqual(simplifyDateTimePatternToParts('X'), [
    { type: 'timeZoneName', value: 'zz' }
  ])
  deepStrictEqual(simplifyDateTimePatternToParts('x'), [
    { type: 'timeZoneName', value: 'zz' }
  ])
})

test('uppercase', () => {
  deepStrictEqual(
    simplifyDateTimePatternToParts('G E y M d H m s S a z', { letterCase: 'uppercase' }), [
      { type: 'era', value: 'EE' },
      { type: 'literal', value: ' ' },
      { type: 'weekday', value: 'WWW' },
      { type: 'literal', value: ' ' },
      { type: 'year', value: 'YYYY' },
      { type: 'literal', value: ' ' },
      { type: 'month', value: 'MM' },
      { type: 'literal', value: ' ' },
      { type: 'day', value: 'DD' },
      { type: 'literal', value: ' ' },
      { type: 'hour', value: 'HH' },
      { type: 'literal', value: ' ' },
      { type: 'minute', value: 'MM' },
      { type: 'literal', value: ' ' },
      { type: 'second', value: 'SS' },
      { type: 'literal', value: ' ' },
      { type: 'fractionalSecond', value: 'MMM' },
      { type: 'literal', value: ' ' },
      { type: 'dayPeriod', value: 'AA' },
      { type: 'literal', value: ' ' },
      { type: 'timeZoneName', value: 'ZZ' }
  ])
})


test/types.test.ts
import { simplifyDateTimePattern, simplifyDateTimePatternToParts, PatternPart } from '../lib/index.js'

type testCallback = () => void
declare function test (label: string, callback: testCallback): void

test('Type declarations for TypeScript', () => {
  let _placeholder: string = simplifyDateTimePattern('G')
  _placeholder = simplifyDateTimePattern('G', {})
  _placeholder = simplifyDateTimePattern('G', { letterCase: 'uppercase' })
  let _parts: PatternPart[] = simplifyDateTimePatternToParts('G')
  _parts = simplifyDateTimePatternToParts('G', {})
  _parts = simplifyDateTimePatternToParts('G', { letterCase: 'uppercase' })
})


test/errors.test.js
import tehanu from 'tehanu'
import { throws } from 'assert'
import { simplifyDateTimePattern } from '../lib/index.js'

const test = tehanu(import.meta.url)

test('throws an error if the format string contains an unescaped latin alphabet character', () => {
  throws(simplifyDateTimePattern.bind(null, 'yyyy-MM-dd-nnnn'), RangeError)
})

test('throws an error if the format string contains unbalanced apostrophes', () => {
  throws(simplifyDateTimePattern.bind(null, "'yyyy"), SyntaxError)
})


lib/index.js
// Placeholder tokens

const ee = { type: 'era', value: 'ee' }
const yy = { type: 'year', value: 'yy' }
const yyyy = { type: 'year', value: 'yyyy' }
const mm1 = { type: 'month', value: 'mm' }
const mmm1 = { type: 'month', value: 'mmm' }
const www = { type: 'weekday', value: 'www' }
const dd = { type: 'day', value: 'dd' }
const hh = { type: 'hour', value: 'hh' }
const mm2 = { type: 'minute', value: 'mm' }
const ss = { type: 'second', value: 'ss' }
const mmm2 = { type: 'fractionalSecond', value: 'mmm' }
const aa = { type: 'dayPeriod', value: 'aa' }
const zz = { type: 'timeZoneName', value: 'zz' }

// Formatting Token replacements

const tokens = new Map(Object.entries({
  G: ee,
  GG: ee,
  GGG: ee,
  GGGG: ee,
  GGGGG: ee,
  y: yyyy,
  yy: yy,
  yyy: yyyy,
  yyyy: yyyy,
  yyyyy: yyyy,
  M: mm1,
  MM: mm1,
  MMM: mmm1,
  MMMM: mmm1,
  MMMMM: mmm1,
  E: www,
  EE: www,
  EEE: www,
  EEEE: www,
  EEEEE: www,
  d: dd,
  dd: dd,
  h: hh,
  hh: hh,
  H: hh,
  HH: hh,
  K: hh,
  KK: hh,
  k: hh,
  kk: hh,
  m: mm2,
  mm: mm2,
  s: ss,
  ss: ss,
  S: mmm2,
  SS: mmm2,
  SSS: mmm2,
  SSSS: mmm2,
  a: aa,
  aa: aa,
  aaa: aa,
  aaaa: aa,
  aaaaa: aa,
  b: aa,
  bb: aa,
  bbb: aa,
  bbbb: aa,
  bbbbb: aa,
  B: aa,
  BB: aa,
  BBB: aa,
  BBBB: aa,
  BBBBB: aa,
  Z: zz,
  ZZ: zz,
  ZZZ: zz,
  ZZZZ: zz,
  ZZZZZ: zz,
  z: zz,
  zz: zz,
  zzz: zz,
  zzzz: zz,
  O: zz,
  OOOO: zz,
  X: zz,
  XX: zz,
  XXX: zz,
  XXXX: zz,
  XXXXX: zz,
  x: zz,
  xx: zz,
  xxx: zz,
  xxxx: zz,
  xxxxx: zz
}))

// Pattern simplifier

function pushLiteral(parts, value) {
  const { length } = parts
  if (length > 0) {
    const last = parts[length - 1]
    if (last.type === 'literal') {
      last.value += value
      return
    }
  }
  parts.push({ type: 'literal', value })
}

export function simplifyDateTimePatternToParts(pattern, options) {
  const parts = []
  for (let i = 0, l = pattern.length; i < l;) {
    let char = pattern[i]
    // Letters are formatting tokens
    if (char >= 'A' && char <= 'Z' || char >= 'a' && char <= 'z') {
      // Multiple occurrences of the same letter are a single token
      const start = i
      do {
        char = pattern[++i]
      } while (char >= 'A' && char <= 'Z' || char >= 'a' && char <= 'z')
      const token = pattern.substring(start, i)
      let part = tokens.get(token)
      if (!part) throw new RangeError(`invalid token: "${token}"`)
      if (options?.letterCase === 'uppercase') {
        const { type, value } = part
        part = { type, value: value.toUpperCase() }
      }
      parts.push(part)
    } else if (char === '\'') {
      // Apostrophes surround a literal to be formatted as-is
      const start = i + 1
      for(;;) {
        if (++i === l) throw new SyntaxError('missing trailing "\'"')
        char = pattern[i]
        if (char === '\'') {
          // An apostrophe can be escaped by doubling
          if (pattern[i + 1] === '\'') {
            ++i
            continue
          }
          break
        }
      }
      // An apostrophe can be escaped by doubling
      const literal = start === i ? '\'' : pattern.substring(start, i).replaceAll('\'\'', '\'')
      pushLiteral(parts, literal)
      ++i
    } else {
      // Punctuation is formatted as-is
      const start = i
      do {
        if (++i === l) break
        char = pattern[i]
      } while (!(char >= 'A' && char <= 'Z' || char >= 'a' && char <= 'z' || char === '\''))
      const punctuation = pattern.substring(start, i)
      pushLiteral(parts, punctuation)
    }
  }

  return parts
}

export function simplifyDateTimePattern(pattern, options) {
  const parts = simplifyDateTimePatternToParts(pattern)
  let result = ''
  for (const { value } of parts) {
    result += value
  }
  if (options?.letterCase === 'uppercase') result = result.toUpperCase()
  return result
}


lib/index.d.ts
/**
 * An object describing a part of a pattern.
 */
export interface PatternPart {
  /**
   * A type of the part compatible with [Intl.DateTiemFormat:formatToParts](
   * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/formatToParts).
   */
  type: 'era' | 'year' | 'month' | 'day' | 'weekday' | 'hour' | 'minute' |
        'second' | 'fractionalSecond' | 'dayPeriod'| 'timeZoneName' | 'literal'
  /**
   * The textual part of the pattern.
   */
  value: string
}

/**
 * An object with the options to control the pattern simplification.
 */
export interface SimplifyOptions {
  /**
   * If the leters in the simplified pattern should be all lowercase or uppercase.
   * @default 'lowercase'
   */
  letterCase?: 'lowercase' | 'uppercase'
}

/**
 * Returns a simplified pattern to show to the end-user in a date/time picker
 * as an array of parts, which values are supposed to be concatenated together.
 *
 * ### Pattern Conversion
 * 
 * The characters wrapped between two single quotes characters (') are escaped.
 * Two single quotes in a row, whether inside or outside a quoted sequence,
 * represent a "real" single quote. (see the last example)
 *
 * The input pattern is based on [Unicode Technical Standard #35](
 * https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table).
 * Only those tokens are supported, which can appear in patterns for formatting
 * instances of `Date` by `Intl.DateTimeFormat`. For example, only the "formatting"
 * variant of a token (M, E) is supported, not the "stand-alone" one (L, i).
 * ("Formatting" means declined according to the rules of the language
 *  in the context of a date. "Stand-alone" mean s always nominative singular.)
 * 
 * Accepted patterns:
 * 
 * | Unit                         | Pattern | Placeholder | Meaning      |
 * |------------------------------|---------|-------------|--------------|
 * | Era                          | G...    | ee          | era          |
 * | Calendar year                | y       | yyyy        | full year    |
 * |                              | yy      | yy          | 2-digit year |
 * |                              | yyy...  | yyyy        | full year    |
 * | Month                        | M | MM  | mm          | month number |
 * |                              | MMM...  | mmm         | month name   |
 * | Day of month                 | d | dd  | dd          | day number   |
 * | Day of week                  | E...    | www         | weekday name |
 * | AM, PM                       | a...    | aa          | AM/PM        |
 * | AM, PM, noon, midnight       | b...    | aa          |              |
 * | Flexible day period          | B...    | aa          |              |
 * | Hour [1-12]                  | h | hh  | hh          | hours        |
 * | Hour [0-23]                  | H | HH  | hh          |              |
 * | Hour [0-11]                  | K | KK  | hh          |              |
 * | Hour [1-24]                  | k | kk  | hh          |              |
 * | Minute                       | m | mm  | mm          | minutes      |
 * | Second                       | s | ss  | ss          | seconds      |
 * | Fraction of second           | S...    | mmm         | milliseconds |
 * | Timezone (ISO-8601 w/ Z)     | X...    | zz          | time zone    |
 * | Timezone (ISO-8601 w/o Z)    | x...    | zz          |              |
 * | Timezone (GMT)               | O...    | zz          |              |
 * | Timezone (specific non-loc.) | z...    | zz          |              |
 * 
 * Although some placeholders are the same, their meaning is clear from the context
 * within the pattern, because people know how to write down a date/time value.
 * 
 * @param pattern a date/time pattern consisting of Unicode LDML tokens
 * @returns an array of parts of the simplified date/time value placeholder
 * @throws {RangeError} if format string contains unbalanced apostrophes (single quotes)
 * @throws {SyntaxError} if format string contains an invalid letter (Unicode LDML token)
 *
 * @example
 * // Represent 11 February 2014 in middle-endian format
 * simplifyDateTimePatternToParts('MM/dd/yyyy')
 * //=> [{ type: 'month', value: 'mm',
 *         type: 'literal', value: '/',
 *         type; 'day', value: 'dd',
 *         type: 'literal', value: '/',
 *         type: 'year': value: 'yyyy' }]
 *
 * @example
 * // Escape string by single quote characters
 * simplifyDateTimePatternToParts("h 'o''clock'")
 * //=> [{ type: 'hour', value: 'hh',
 *         type: 'literal', value: ' o\'clock' }]
 */
export function simplifyDateTimePatternToParts(pattern: string, options?: SimplifyOptions): PatternPart[]

/**
 * Returns a simplified pattern to show to the end-user in a date/time picker.
 *
 * ### Pattern Conversion
 * 
 * The characters wrapped between two single quotes characters (') are escaped.
 * Two single quotes in a row, whether inside or outside a quoted sequence,
 * represent a "real" single quote. (see the last example)
 *
 * The input pattern is based on [Unicode Technical Standard #35](
 * https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table).
 * Only those tokens are supported, which can appear in patterns for formatting
 * instances of `Date` by `Intl.DateTimeFormat`. For example, only the "formatting"
 * variant of a token (M, E) is supported, not the "stand-alone" one (L, i).
 * ("Formatting" means declined according to the rules of the language
 *  in the context of a date. "Stand-alone" mean s always nominative singular.)
 * 
 * Accepted patterns:
 * 
 * | Unit                         | Pattern | Placeholder | Meaning      |
 * |------------------------------|---------|-------------|--------------|
 * | Era                          | G...    | ee          | era          |
 * | Calendar year                | y       | yyyy        | full year    |
 * |                              | yy      | yy          | 2-digit year |
 * |                              | yyy...  | yyyy        | full year    |
 * | Month                        | M | MM  | mm          | month number |
 * |                              | MMM...  | mmm         | month name   |
 * | Day of month                 | d | dd  | dd          | day number   |
 * | Day of week                  | E...    | www         | weekday name |
 * | AM, PM                       | a...    | aa          | AM/PM        |
 * | AM, PM, noon, midnight       | b...    | aa          |              |
 * | Flexible day period          | B...    | aa          |              |
 * | Hour [1-12]                  | h | hh  | hh          | hours        |
 * | Hour [0-23]                  | H | HH  | hh          |              |
 * | Hour [0-11]                  | K | KK  | hh          |              |
 * | Hour [1-24]                  | k | kk  | hh          |              |
 * | Minute                       | m | mm  | mm          | minutes      |
 * | Second                       | s | ss  | ss          | seconds      |
 * | Fraction of second           | S...    | mmm         | milliseconds |
 * | Timezone (ISO-8601 w/ Z)     | X...    | zz          | time zone    |
 * | Timezone (ISO-8601 w/o Z)    | x...    | zz          |              |
 * | Timezone (GMT)               | O...    | zz          |              |
 * | Timezone (specific non-loc.) | z...    | zz          |              |
 * 
 * Although some placeholders are the same, their meaning is clear from the context
 * within the pattern, because people know how to write down a date/time value.
 * 
 * @param pattern a date/time pattern consisting of Unicode LDML tokens
 * @returns a string with the simplified date/time value placeholder
 * @throws {RangeError} if format string contains unbalanced apostrophes (single quotes)
 * @throws {SyntaxError} if format string contains an invalid letter (Unicode LDML token)
 *
 * @example
 * // Represent 11 February 2014 in middle-endian format
 * simplifyDateTimePattern('MM/dd/yyyy')
 * //=> 'mm/dd/yyyy'
 *
 * @example
 * // Escape string by single quote characters
 * simplifyDateTimePattern("h 'o''clock'")
 * //=> "hh o'clock"
 */
export function simplifyDateTimePattern(pattern: string, options?: SimplifyOptions): string
